### ISSUE TITLE
Title: Live streams - can't stop download

### ISSUE THREAD

**Issue by Spirograph12** [mhogomchungu/media-downloader#236]
OS: Ubuntu 20.04.5 LTS
media-downloader_2.7.0-1_amd64.deb
yt-dlp 2022.11.11
ffmpeg 4.2.7-0ubuntu0.1

Media Downloader is an excellent program, very flexible. Good work! But it suffers from a familiar live stream bug: yt-dlp uses ffmpeg to download live streams. Here are examples of live streams:

https://www.cbsnews.com/live/

https://wnetdt.lls.pbs.org/out/v1/0456457548354b88b32fc437e4e7ee01/index.m3u8

https://classicarts.akamaized.net/hls/live/1024257/CAS/master.m3u8

https://playoutengine.sinclairstoryline.com/playout/9f87522c-5a0e-4ff4-b82c-d5564216132f.m3u8

https://docubayvh.s.llnwi.net/526a07ab-6ae7-4b6c-84a1-159791416484_1000004372_HLS/manifest.m3u8

To download live streams using yt-dlp, you use these flags:

`--downloader ffmpeg --hls-use-mpegts --no-part`

And also you set output template to something like this:

`%(title)s %(id)s.ts`

Now live streams will be correctly saved as mpeg transport streams with extension .ts

But in Media Downloader, once you start downloading a live stream, you cannot stop it. Hitting Cancel only kills yt-dlp, but not the child process ffmpeg. Therefore, the end user is forced to stop ffmpeg manually, using a kill command or task manager. This is not ideal.

Ideally, Media Downloader should detect when yt-dlp has spawned an ffmpeg child process, such as downloading a live stream. When the user hits Cancel, a kill signal should be sent to both yt-dlp and ffmpeg.

There is probably some elegant way to do this. But lacking an elegant solution, at least add a "Stop ffmpeg" button to the GUI. You could give it a tooltip like "Stops download of live streams by ffmpeg".

For an example of how a different GUI handled these issues, please see the old yt-dlg 0.4 here:

https://github.com/MrS0m30n3/youtube-dl-gui

This Ubuntu binary still works well on Ubuntu 20.04 (or Linux Lite 5.8):

http://ppa.launchpad.net/nilarimogard/webupd8/ubuntu/pool/main/y/youtube-dlg/youtube-dlg_0.4-1~webupd8~disco0_all.deb

There, the GUI can handle multiple live streams simultaneously. The GUI keeps track of every ffmpeg thread spawned by youtube-dl. Individual live streams can be stopped and others added or restarted. This is true of the Linux version. The Windows version apparently has trouble stopping ffmpeg, as discussed here:

https://github.com/MrS0m30n3/youtube-dl-gui/issues/49#issuecomment-121231385

Whatever approach you take, the ability to easily stop download of live streams would make Media Downloader a more powerful tool for end users.

Thanks very much for considering my suggestions.

**Comment by mhogomchungu** [comment]
Done thanks to [this](https://www.baeldung.com/linux/get-process-child-processes) post.

I converted the script on that page to below code abd it seems to work.

https://github.com/mhogomchungu/media-downloader/blob/1eb0d12907af5f914b8941862ae75f7e8eb379c6/src/utility.cpp#L412-L436

**Comment by Spirograph12** [comment]
Thanks for your quick response! I'm not a build-from-source guy, but I decided to give it a try. On Ubuntu 20.04, the development version of QT libraries seems to be called qtcreator:

https://linuxways.net/ubuntu/how-to-install-qt-on-ubuntu-20-04/

I followed your steps and the build went fine. In light testing, your mod seems to do the trick! Hitting Cancel now stops ffmpeg downloading the live stream. Bravo!
